### ISSUE TITLE
VS2019 - Python installation directory is not consistent

### ISSUE THREAD

**Issue by vsfeedback** [microsoft/PTVS#7836]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/VS2019---Python-installation-directory-i/10581676)._

---
I am responsible for installing Visual Studio 2019 on a number of CI build machines.

Using the Visual Studio installer I install the Python development workload and then Python 3 64-bit (3.9.13) and Python 3 64-bit (3.7.8).

After the installation is complete I check where the Python 3.7.8 python.exe binary was installed and it’s one of these:

*   %LOCALAPPDATA%\\Programs\\Python\\Python37
*   C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python37\_64

Most of the time it’s the %LOCALAPPDATA% location, but not always. I have no idea when the installer will choose one directory over the other. We need to know the actual installation directory because we reference it in our build scripts.

Thanks!

The TLDR, if you don't want to read all of the replication details below:
- When you start with a brand new machine and install VS2019 with Python 3.7 it will install Python in a global location (C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64). Even if you uninstall and reinstall Python it will still be installed in the global location.
- If you install VS Python using a different user it will confuse VS Installer but a repair will fix it - and will install Python in a per-user location (C:\Users\username\AppData\Local\Programs\Python\Python37).
- Once you're in this state it’s not possible to go back to installing to the global location, even if uninstall everything, including using InstallCleanup.exe. There is some state that forces the installer to use the per-user location and that state doesn’t get cleaned up via the documented uninstall methods. I would really like to know where the state is so I can install VS Python into the global location.

The workaround is to always use the same user to make changes to Visual Studio.  But that doesn't help the machines that are stuck in the per-user state.

---
#### Eric Pederson on 2/8/2024, 09:33 AM: 

(private comment, text removed)
#### Eric Pederson on 2/8/2024, 09:38 AM: 

(private comment, text removed)
#### Feedback Bot on 2/8/2024, 05:36 PM: 

(private comment, text removed)
#### Eric Pederson on 2/8/2024, 05:39 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by StellaHuang95** [comment]
Hi there, thanks for reporting the issues. Unfortunately, we've seen similar issues with VS installer in the past and it's due to how the python windows installer works. So the workaround is not to install the python interpreter through the VS installer. The good news is that we've marked this component as out of support moving forward (for future VS releases), so it should always be unchecked by default. Please try downloading the Python runtime directly from https://python.org/ and install it manually. Feel free the reopen this issue if that doesn't solve your problem. Thanks!

**Comment by sourcedelica** [comment]
I ended up finding a way to get back to the state where I could install Python using the VS Installer to the global location (`C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64`).  

The steps I used were:

1. Uninstall Python 3.7.8 and the Python workload using the VS Installer
2. Uninstall Python 3.7.8 using `appwiz.cpl` (old-school Add/Remove Programs)
3. Using Registry Editor, search for `python` and delete everything that looked like Python 3.7.

I had done #1 and #2 before with no luck.  The registry surgery was needed to get back to the original state.
